hubble: Add support for SockLB tracing

gandro opened the pull request:
This PR adds a Hubble parser for `TraceSockNotify` events. Those
events are emitted whenever SockLB is performed, see cilium/cilium#20492
for details about the datapath tracing.

This parser is similar to the L3/L4 event parser. But because the events
are traced on the socket level, we do not know the source endpoint id,
source endpoint IP address or source TCP/UDP port of the event. We only
know the source cgroup ID. From the cgroup, we can derive which local
pod emitted the event (via `GetPodMetadataForContainer`). From the pod
IP we can then perform an IP-based lookup in either the local endpoint
list or the IPCache to determine the source endpoint metadata (endpoint
id, numeric identity, labels, etc). The lookup for destination IP is
almost identical to the threefour parser, with the exception that we do
not have any datapath identity available for fallback.

The resulting flow looks similar to regular `FlowType_L3_L4`, with a
couple of key differences:

 - The cgroup id of the container is available
 - Information if the event is pre or post translation point is
   available
 - The socket cookie is available
 - The TCP/UDP source port is _not_ known
 - TCP flags are _not_ known
 - Ethernet headers are _not_ known

Unfortunately, sock events can also be performed by non-pod entities
(e.g. the host cgroup). In that case, we cannot perform a mapping to any
local endpoint and therefore the source IP and endpoint data are left
empty. The cgroup id may also not be available on older kernels or
systems with the systemd cgroup driver (though this will likely be fixed
in a follow-up PR). This fact that the source IP may be empty/unknown is
also key difference to the flows which have been produced by Hubble so
far.

---

Here's an example of a pod performing a DNS request and response. The first three events are the DNS request: We first see two the SockLB events (pre and post translation, meaning before and after service backend selection):

![image](https://user-images.githubusercontent.com/50564/195646798-37826ca5-18f7-45f6-a748-6861d1e16a42.png)

And here are the events for the response: We see the response packet arriving at the endpoint, but before it is delivered over the socket, reverse NAT being applied, i.e. the pod IP is once again translated to the service IP (which is where the userspace sent the packet to):

![image](https://user-images.githubusercontent.com/50564/195647735-57dfc515-af1d-4db2-a91d-32cbd1a83a76.png)

gandro committed:
api: Update Hubble API to support TraceSockNotify

This adds the relevant fields, including the translation point, socket
cookie and cgroup id. The cgroup id allocated by the kernel can never be
zero, so using zero as the "absent" value is possible here.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
hubble: Add support for SockLB tracing

This commit adds a Hubble parser for `TraceSockNotify` events. Those
events are emitted whenever SockLB is performed, see cilium/cilium#20492
for details about the datapath tracing.

This parser is similar to the L3/L4 event parser. But because the events
are traced on the socket level, we do not know the source endpoint id,
source endpoint IP address or source TCP/UDP port of the event. We only
know the source cgroup ID. From the cgroup, we can derive which local
pod emitted the event (via `GetPodMetadataForContainer`). From the pod
IP we can then perform an IP-based lookup in either the local endpoint
list or the IPCache to determine the source endpoint metadata (endpoint
id, numeric identity, labels, etc). The lookup for destination IP is
almost identical to the threefour parser, with the exception that we do
not have any datapath identity available for fallback.

The resulting flow looks similar to regular `FlowType_L3_L4`, with a
couple of key differences:

 - The cgroup id of the container is available
 - Information if the event is pre or post translation point is
   available
 - The socket cookie is available
 - The TCP/UDP source port is _not_ known
 - TCP flags are _not_ known
 - Ethernet headers are _not_ known

Unfortunately, sock events can also be performed by non-pod entities
(e.g. the host cgroup). In that case, we cannot perform a mapping to any
local endpoint and therefore the source IP and endpoint data are left
empty. The cgroup id may also not be available on older kernels or
systems with the systemd cgroup driver (though this will likely be fixed
in a follow-up PR). This fact that the source IP may be empty/unknown is
also key difference to the flows which have been produced by Hubble so
far.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
hubble/parser: Extract functions to be shared among parsers

This commit moves out useful functions out of the threefour parser
package into a common package to be shared among parsers.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
hubble/parser: Extract resolveEndpoint helper

This allows us to call this helper function from other parsers as well.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

aditighag commented on pkg/hubble/parser/sock/parser.go:
```diff
@@ -0,0 +1,221 @@
+		srcEndpoint, dstEndpoint = dstEndpoint, srcEndpoint
+	}
+
+	decoded.Verdict = flowpb.Verdict_FORWARDED
```
Can we remove the hard-coded forwarded verdict? This policy related field isn't relevant for socket-lb tracing events.

gandro replied:
I'm not sure if the trace sock events are fundamentally different from e.g. trace events emitted by FROM_LXC. Those also are also emitted way before policy is applied and contain the same verdict. 

In general, I agree with your sentiment, but in the context of the other Hubble events we have, I do think FORWARDED makes somewhat sense here. We could add a TRANSLATED verdict, maybe?

gandro replied:
Discussed with the Hubble team, we'll add a new verdict kind for those events.

gandro replied:
Added the `TRANSLATED` verdict

aditighag commented on pkg/hubble/parser/sock/parser.go:
```diff
@@ -0,0 +1,221 @@
+	decoded.L4 = decodeL4(sock.L4Proto, srcPort, dstPort)
+	decoded.Source = srcEndpoint
+	decoded.SourceNames = p.resolveNames(dstEndpoint.GetID(), srcIP)
+	decoded.SourceService = p.decodeService(srcIP, srcPort)
```
What's `SourceService` field for?

gandro replied:
Note that we swap source and destination for rev nat events. This therefore can be the service name of the "DstIP". 
Check the third flow of the second screenshot above, the `default/kube-dns:53` service name is what's being populated here.

aditighag commented on pkg/hubble/parser/sock/parser.go:
```diff
@@ -0,0 +1,221 @@
+}
+
+func (p *Parser) decodeSrcIP(cgroupId uint64, ipVersion flowpb.IPVersion) net.IP {
+	if p.cgroupGetter != nil {
```
When would `p.cgroupGetter` be `nil`? We should fail more assertively instead of having a check for every trace event if it's `nil`.

gandro replied:
It's only nil in unit testing. Other Hubble code does the same. The reason being that we've had bug reports of OSS Fuzz: https://github.com/cilium/cilium/pull/20446

Arguably a bug in their framework, but not requiring this to be set simplifies fuzzing nonetheless.

aditighag replied:
I expressed the same concern as Joe did on the PR you linked - https://github.com/cilium/cilium/pull/20446#discussion_r918669500. 
It doesn't look like we followed up doing `nil` checks in constructors that Tobias proposed, did we?

gandro replied:
No, we did not. But this is a discussion that goes beyond this PR, so I'll keep this as is right now. 

Let's have the Hubble code handle this consistently in all packages. I don't want the socklb parser to do it differently than the other parsers do.

aditighag commented on pkg/hubble/parser/sock/parser.go:
```diff
@@ -0,0 +1,221 @@
+
+func (p *Parser) decodeSrcIP(cgroupId uint64, ipVersion flowpb.IPVersion) net.IP {
+	if p.cgroupGetter != nil {
+		if m := p.cgroupGetter.GetPodMetadataForContainer(cgroupId); m != nil {
```
Hmm... We should prune calls from host processes (essentially non-pods). Host cgroup id will usually be `1`, except for kind like nested environments. Although there can also be events initiated from non-pod entities like `kubelet` which can have its own cgroup id. 

Maybe datapath can assist here by indicating whether events are initiated from the host network namespace? We already have logic baked into the datapath that identified "host" network namespace on nested environments like kind.

gandro replied:
Hm. Hubble so far does not do any filtering of events by default, it forwards everything that it receives and asks the user to filter out irrelevant events.

I would therefore also prefer to filter this in the datapath. If that's not possible and we need to teach Hubble to drop events here, then we need to discuss how to best do this. But I do prefer for `hubble observe` and `cilium monitor` to always display the same events (which is the case right now).

gandro replied:
Discussed with the Hubble team: We'll filter out any events without a matching pod in Hubble, since those events don't provide much value to end users. There will be an agent option to disable this filtering for developers and troubleshooting (but the raw data should also be available via cilium monitor).

gandro replied:
Added a flag to filter them out (see last commit).

aditighag commented on pkg/hubble/parser/sock/parser.go:
```diff
@@ -0,0 +1,221 @@
+
+func (p *Parser) resolveNames(epID uint32, ip net.IP) (names []string) {
+	if p.dnsGetter != nil {
+		return p.dnsGetter.GetNamesOf(epID, ip)
```
Fancy!

aditighag commented on pkg/hubble/parser/sock/parser.go:
```diff
@@ -0,0 +1,221 @@
+	}
+}
+
+func decodeSummary(sock *monitor.TraceSockNotify) string {
```
Summary is printing only the L4 protocol. Is this code complete?

gandro replied:
Yes. The summary refers to the string in parentheses in the following screenshot (e.g. `ICMPv6 RouterSolicitation`)

![image](https://user-images.githubusercontent.com/50564/197740190-78831bd9-5fba-4d0f-9cda-0fa24c5ac884.png)

Since we don't have more information on the actual contents of the packet for trace sock events (e.g. TCP flags), just displaying the protocol seems to be the most consistent with the existing output. But I'm open to suggestions.

kaworu replied:
`Summary` is marked as deprecated, and `decodeSummary` only provides info that we already know through `L4`. IMHO this logic belongs to the client, but I don't feel too strongly about it.

aditighag reviewed (changes requested):
🚀 I only looked at the commit "hubble: Add support for SockLB tracing" that pertains to parsing socket-lb trace events. The parsing logic looks good. Let's follow up on pruning events from non-pod entities. 

Can we print cgroup id like metadata in the verbose json output? Will you be adding json summary in a follow-up?

kaworu commented on pkg/hubble/parser/common/labels.go:
```diff
@@ -0,0 +1,63 @@
+			}
+		} else {
+			filteredLabels = append(filteredLabels, label)
+		}
```
nit: maybe move this branch as the first one in the for loop to reduce indentation, i.e. something like

```go
for _, label := range labels {
    if !strings.HasPrefix(label, cidrPrefix) {
        filteredLabels = append(filteredLabels, label)
        continue
    }
    // max cidr label logic here
}
```

kaworu commented on pkg/hubble/parser/common/endpoint.go:
```diff
@@ -0,0 +1,115 @@
+	resolveIdentityConflict := func(identity identity.NumericIdentity) uint32 {
+		// if the datapath did not provide an identity (e.g. FROM_LXC trace
+		// points), use what we have in the user-space cache
+		userspaceSecurityIdentity := uint32(identity)
```
nit
```suggestion
		userspaceSecurityIdentity := identity.Uint32()
```

kaworu commented on pkg/hubble/parser/common/labels.go:
```diff
@@ -0,0 +1,63 @@
+			if currMask > maxMask {
+				max = curr
+				maxStr = label
+			}
```
nit: IIUC technically we don't need to remember the actual max `IPNet` here, only its mask size. Assuming `var max *net.IPNet` was replaced by `var maxSize int` earlier we could

```suggestion
			if currMask, _ := curr.Mask.Size(); currMask > maxSize {
				maxSize, maxStr = currMask, label
			}
```

This simplify the logic a bit and also avoid recomputing the max mask size.

kaworu commented on pkg/hubble/parser/parser.go:
```diff
@@ -94,7 +103,11 @@ func (p *Parser) Decode(monitorEvent *observerTypes.MonitorEvent) (*v1.Event, er
+		case monitorAPI.MessageTypeDebug:
 			// Debug is currently the only perf ring buffer event without any
 			// associated captured network packet header, so we treat it as
 			// a special case
```
Seems to me that this comment isn't accurate anymore after this patch as we can get [a TraceSock event on connect](https://github.com/cilium/cilium/blob/7cc3e3a3e2f60aba31d4985cb986c98bfc4dae34/bpf/bpf_sock.c#L464) correct?

kaworu commented on pkg/hubble/parser/sock/parser.go:
```diff
@@ -0,0 +1,221 @@
+		return fmt.Errorf("failed to parse sock trace event: %w", err)
+	}
+
+	isRevNat := decodeRevNat(sock.XlatePoint)
```
At first I thought `decodeRevNat` returning a `*wrapperspb.BoolValue` was a left-over from a previous version of the patch, because `isRevNat` is never added to the decoded flow and this line is the only call site. We check for `isRevNat.GetValue()` a few lines below, but we only need a `true`/`false` signal at that point.

Looking at the screenshots of the Hubble CLI output, I noticed that we're missing the arrows for `TraceSock` flows (i.e. we see `<>`), so related question: would setting `IsReply` and `TrafficDirection` based on `isRevNat` make sense for `TraceSock` events?

gandro replied:
Yes, your suspicion is correct, an earlier version of the PR used the result of  `decodeRevNat` as the reply flag. I've changed it to a `bool` in the most recent version.

Unfortunately, trying to derive the traffic direction / reply state from the observation point does not work: `IsReply` cannot be meaningfully populated for TraceSock events because their occur on the socket granularity level, not a packet level. 

The screenshot in the PR description only shows the traces from the node where `pod-to-a` is hosted, where rev nat is indeed only used for the reply. But if we include also the events emitted by the kube-dns host, then the picture looks slightly different, we also see rev nat being used for the non-reply packet:

![image](https://user-images.githubusercontent.com/50564/199265945-4f66fa6d-909f-43eb-b6b3-83c2b0c16b34.png)

The screenshot above is using `xwing`, but it's a single UDP request packet being sent from xwing to kubedns, with a single reply packet being sent back. Please note the timetsamps - as you can see, `kind-worker2` (where coredns is hostd) emits a `pre-xlate-rev` for the initial incoming packet (there is no post event, as there is no translation). The reply packet (the last 4 events) happens half a second later, and does not emit any SockLB event, probably because it's sent back on the same "connected UDP socket".

gandro replied:
Another way to formulate this: I think (but I might be mistaken, cc @aditighag) that `pre/post-xlate-rev` is only executed on `recvmsg` and `getpeername`, both system calls don't really imply anything of overall the "traffic direction" (e.g. if it was an ingress or egress connection in policy lingo).

`pre/post-xlate-fwd` on the other hand is executed for `connect` and `sendmsg`. `connect` is typically used to initiate an egress connection, but the same cannot necessary be said for `sendmsg`.

kaworu replied:
@gandro ah that make sense, thanks for the detailed explanation :+1: one more question about the example:

> The reply packet (the last 4 events) happens half a second later, and does not emit any SockLB event, probably because it's sent back on the same "connected UDP socket".

To make sure I understand correctly: the reply packet does not emit any SockLB event on `kind-worker2` (i.e. `coredns` host), but it does in `kind-worker` (i.e. `xwing` host) correct? The screenshot's last two events are a couple of `pre/post-xlate-rev` analogous to the request packet `pre/post-xlate-fwd` events (which make sense to me), hence the question.

gandro replied:
That's correct, yes. My guess here is that `coredns` is using Golang's `net.Conn` to send the reply packet, meaning it's using the `write` system call on a "connected UDP socket". Thus, none of the above cgroup hooks is hit. `dig` (used inside `xwing`) on the other hand might be using `readmsg`/`sendmsg` syscalls, thus causing a Hubble event for each syscall.

Maybe @aditighag has deeper insight here.

gandro replied:
On second thought, connected UDP sockets are only useful for clients. Checking a basic UDP echo server, it seems Go is using `recvfrom` and `sendto`, so not sure why the `sendto` does not cause a Hubble event (it should for unconnected udp sockets, according to https://manpages.debian.org/experimental/bpftool/bpftool-cgroup.8.en.html)

aditighag replied:
The recvmsg hook is executed only for regular UDP. TCP (and connected UDP) sockets are connected, so TCP clients send and receive data using `send`/`recv` socket calls.
@gandro Your observation is correct. Golang uses a connected UDP client by default. Regarding, `sendto` for connected UDP, here is a snippet from the man page - 

`If sendto() is used on a connection-mode (SOCK_STREAM, SOCK_SEQPACKET) socket, the arguments dest_addr and addrlen are ignored (and the error EISCONN may be returned when they are not NULL and 0)`

aditighag replied:
>Another way to formulate this: I think (but I might be mistaken, cc @aditighag) that pre/post-xlate-rev is only executed on recvmsg and getpeername, both system calls don't really imply anything of overall the "traffic direction" (e.g. if it was an ingress or egress connection in policy lingo).

When podA on nodeA talks to podB on nodeB, both fwd and rev translation happen on the nodeA (for UDP). For TCP and connected UDP, it would just be fwd translation.  

The {pre, post}-fwd/rev markers already convey the direction. We shouldn't try to interpret this with the same context as policy terminology (ingress/egress).

fwd: svc vip -> backend ip
rev: backend ip -> svc vip

Hope that makes sense.

kaworu reviewed (commented):
Awesome work @gandro! I have a few comments, mostly cosmetic. Feel free to ignore nits and comments on code that has only moved (I only noticed after review).

gandro committed:
hubble: Code cleanup in parser/common/endpoint.go

Suggested-by: Alexandre Perrin <alex@isovalent.com>
Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
hubble: Code cleanup in parser/common/labels.go

Suggested-by: Alexandre Perrin <alex@isovalent.com>
Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
hubble: Code simplification in parser/common/labels.go

Suggested-by: Alexandre Perrin <alex@isovalent.com>
Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
hubble: Add option to skip events with unknown cgroup ids

This commit adds a new agent flag which instructs Hubble to skip
TraceSock events with unknown cgroup ids, i.e. cgroup ids for which we
are unable to find a matching pod. Those skipped events are not added to
the ring buffer and can only be observed via `cilium monitor`, unless
the flag is disabled.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

jibi reviewed (commented):
Agent changes look good, but I think we are missing the related changes for the Helm charts

gandro committed:
helm: Add hubble.skipUnknownCGroupIDs value

To avoid cluttering the ConfigMap with default values, it is only added
to the ConfigMap if explicitly set (i.e. not is not `nil`).

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

kaworu commented on pkg/hubble/observer/local_observer.go:
```diff
@@ -127,9 +127,13 @@ nextEvent:
+				// since they are not intended for us (e.g. MessageTypeRecCapture)
 				if !parserErrors.IsErrInvalidType(err) {
 					s.log.WithError(err).WithField("event", monitorEvent).Debug("failed to decode payload")
 				}
```
nit: could handle the invalid type case in the `switch` directly

```suggestion
			switch {
			// silently ignore those kind of errors
			case errors.Is(err, parserErrors.ErrUnknownEventType),
				errors.Is(err, parserErrors.ErrEventSkipped),
				// silently ignore perf ring buffer events with invalid types,
				// since they are not intended for us (e.g. MessageTypeRecCapture)
				parserErrors.IsErrInvalidType(err):
			default:
				s.log.WithError(err).WithField("event", monitorEvent).Debug("failed to decode payload")
```

gandro replied:
Good suggestion, applied!

kaworu reviewed (approved):
:100:

aditighag commented on pkg/hubble/parser/sock/parser.go:
```diff
@@ -0,0 +1,217 @@
+		srcEndpoint, dstEndpoint = dstEndpoint, srcEndpoint
+	}
+
+	decoded.Verdict = flowpb.Verdict_TRANSLATED
```
We should set this only on the post_* notifications. This isn't applicable for pre_* events.

gandro replied:
@aditighag PTAL. I've changed the code to use TRANSLATED for post events and the newly introduced TRACED verdict for pre events (as discussed with the Hubble team a week ago, we want a verdict here. Verdict UNKNOWN will be confusing to users, since it indicates missing data):

https://github.com/cilium/cilium/blob/c5372514225a0335457d65fb66fdec8b72cf03b3/pkg/hubble/parser/sock/parser.go#L215-L226

I've also added a short comment to each verdict value to document what it does.

aditighag reviewed (changes requested):
Commit "hubble: Add support for SockLB tracing" looks good to me. Nice work! 

One minor comment around setting the verdict field.

squeed reviewed (approved):
Looks good for my files!

rolinh reviewed (approved):
:shipit:
